app/api/v1/auth/index.js

```diff
@@ -0,0 +1,90 @@
+const express = require('express');
+const jwt = require('jsonwebtoken');
+const config = require('../../../config');
+const error = require('../../../error');
+const { User, Activity } = require('../../../db');
+
+const router = express.Router();
+
+const TOKEN_EXPIRES = 86400; // 1 day in seconds
+
+/**
+ * Middleware function that determines if a user is authenticated and assigns req.user to their user info from the db.
+ * Auth header should be included in the 'Authorization' request header in the form of 'Bearer <TOKEN>'.
+ */
+const authenticated = (req, res, next) => {
+  const authHeader = req.get('Authorization');
+  if (!authHeader) return next(new error.Unauthorized());
+
+  const authHead = authHeader.split(' ');
+  const invalidAuthFormat = authHead.length !== 2 || authHead[0] !== 'Bearer' || authHead[1].length < 1;
+  if (invalidAuthFormat) return next(new error.Unauthorized());
+
+  const token = authHead[1];
+  jwt.verify(token, config.auth.secret, (err, decoded) => {
+    if (err) return next(new error.Unauthorized());
+
+    // if the user provided a valid token, use it to deserialize the UUID to an actual user object
+    User.findByUUID(decoded.uuid).then((user) => {
+      if (!user) throw new error.Unauthorized();
+      req.user = user;
+    }).then(next).catch(next);
+  });
+};
+
+/**
+ * Login route. POST body should be in the format of { email, password } and returns an auth token on success.
+ */
+router.post('/login', (req, res, next) => {
+  if (!req.body.email || req.body.email.length < 1) return next(new error.BadRequest('Email must be provided'));
```

Comment:
is this to account for cases where your email is just an empty '' string? If so should the same logic also apply to the User in the following functions?

Reply:
Not sure I understand. The route accepts an email and password, and each must be some non-empty string. Then, [`User::findByEmail`](https://github.com/acmucsd/membership-portal/blob/master/app/db/models/user.js#L182) queries the database and retrieves the `User` object attached to the given email. Then if the user is in good standing, [`User.prototype::verifyPassword`](https://github.com/acmucsd/membership-portal/blob/master/app/db/models/user.js#L241) checks if the given password matches the hash associated with that `User` object. Then, a JWT will be signed by the app and returned in the response JSON. We validate every piece of data before writing it to the database, so therefore we know that the `User` object we retrieve from the database is well-formed.